hyperspy/events.py

```diff
@@ -0,0 +1,523 @@
+        if nargs not in self._connected:
+            self._connected[nargs] = set()
+        self._connected[nargs].add(function)
+
+    def disconnect(self, function):
+        """
+        Disconnects a function from the event. The passed function will be
+        disconnected irregardless of which 'nargs' argument was passed to
+        connect().
+
+        If you only need to temporarily prevent a function from being called,
+        single callback suppression is supported by the `suppress_callback`
+        context manager.
+
+        See also
+        --------
+        connect
+        suppress_callback
+        """
+        for c in self._connected.itervalues():
+            if function in c:
+                c.remove(function)
+
+    def _trigger_nargs(self, f, args, kwargs, nargs):
+        """
+        Basic trigger resolution.
+        """
+        if len(args) < nargs:
+            # To few args!
+            # We try to fill with ordered kwargs if present.
+            # If order kwargs not present, or not enough, fill with matching
+            # kwargs in the order that function defines them
+            args = list(args)
+            kwargs = kwargs.copy()
+            if self.arguments:
+                i = 0
+                while len(args) < nargs and i < len(self.arguments):
+                    if self.arguments[i] in kwargs:
+                        args.append(kwargs.pop(self.arguments[i]))
+                    i += 1
+            if len(args) < nargs:
+                spec = inspect.getargspec(f)
+                candidates = list(spec.args)[len(args):]
+                if inspect.ismethod(f):
+                    candidates = candidates.pop()     # Remove `self`
+                matches = [kwargs[c] for c in candidates if c in kwargs]
+                args.extend(matches[:nargs-len(args)])
+        return f(*args[0:nargs])
+
+    def trigger(self, *args, **kwargs):
+        """
+        Triggers the event. If the event is suppressed, this does nothing.
+        Otherwise it calls all the connected functions with the arguments as
+        specified when connected.
+
+        See also
+        --------
+        suppress
+        suppress_callback
+        Events.suppress
+        """
+        if not self._suppress:
+            # Loop on copy to deal with callbacks which change connections
+            for nargs, c in self._connected.copy().iteritems():
+                if nargs is 'all':
+                    for f in c:
+                        f(*args, **kwargs)
+                else:
+                    if len(args) + len(kwargs) < nargs:
+                        raise ValueError(
+                            ("Tried to call %s which require %d args " +
+                             "with only %d.") % (str(c), nargs, len(args) +
+                                                 len(kwargs)))
+                    for f in c.copy():
+                        self._trigger_nargs(f, args, kwargs, nargs)
```

Comment:
What if the triggered function modifies an argument? It may be too expensive to copy all of them all the time, however it could lead to strange behaviour, if one of the earlier triggered functions changes one of the passed arguments to the later triggered ones..